src/gui/qgscrsdefinitionwidget.cpp

```diff
@@ -0,0 +1,282 @@
+                                  tr( "This proj projection definition is valid." ) );
+      }
+      else
+      {
+        QMessageBox::warning( this, tr( "Custom Coordinate Reference System" ),
+                              tr( "This proj projection definition is not valid:" ) + QStringLiteral( "\n\n" ) + projErrors.join( '\n' ) );
+      }
+      break;
+    }
+  }
+
+  // reset logger to terminal output
+  proj_log_func( context, nullptr, nullptr );
+  proj_context_destroy( context );
+  context = nullptr;
+}
+
+void QgsCrsDefinitionWidget::formatChanged()
+{
+  QgsCoordinateReferenceSystem crs;
+  QString newFormatString;
+  switch ( static_cast< QgsCoordinateReferenceSystem::Format >( mFormatComboBox->currentData().toInt() ) )
+  {
+    case QgsCoordinateReferenceSystem::FormatProj:
+    {
+      crs.createFromWkt( multiLineWktToSingleLine( mTextEditParameters->toPlainText() ) );
+      if ( crs.isValid() )
+        newFormatString = crs.toProj();
+      break;
+    }
+
+    case QgsCoordinateReferenceSystem::FormatWkt:
+    {
+      PJ_CONTEXT *pjContext = QgsProjContext::get();
+      QString proj = mTextEditParameters->toPlainText();
+      proj.replace( QLatin1String( "+type=crs" ), QString() );
+      proj += QLatin1String( " +type=crs" );
+      QgsProjUtils::proj_pj_unique_ptr crs( proj_create( QgsProjContext::get(), proj.toLatin1().constData() ) );
+      if ( crs )
+      {
+        const QByteArray multiLineOption = QStringLiteral( "MULTILINE=YES" ).toLocal8Bit();
+        const char *const options[] = {multiLineOption.constData(), nullptr};
+        newFormatString = QString( proj_as_wkt( pjContext, crs.get(), PJ_WKT2_2019, options ) );
+      }
+      break;
+    }
+  }
+  if ( !newFormatString.isEmpty() )
+    mTextEditParameters->setPlainText( newFormatString );
+}
+
+void QgsCrsDefinitionWidget::pbnCalculate_clicked()
+{
+  // We must check the prj def is valid!
+  PJ_CONTEXT *pContext = QgsProjContext::get();
+  QString projDef = mTextEditParameters->toPlainText();
+  QgsDebugMsgLevel( QStringLiteral( "Proj: %1" ).arg( projDef ), 3 );
+
+  // Get the WGS84 coordinates
+  bool okN, okE;
+  double latitude = mNorthWGS84Edit->text().toDouble( &okN );
+  double longitude = mEastWGS84Edit->text().toDouble( &okE );
+
+  if ( !okN || !okE )
+  {
+    QMessageBox::warning( this, tr( "Custom Coordinate Reference System" ),
+                          tr( "Northing and Easting must be in decimal form." ) );
+    mProjectedXLabel->clear();
+    mProjectedYLabel->clear();
+    return;
+  }
+
+  if ( static_cast< QgsCoordinateReferenceSystem::Format >( mFormatComboBox->currentData().toInt() ) == QgsCoordinateReferenceSystem::FormatProj )
+    projDef = projDef + ( projDef.contains( QStringLiteral( "+type=crs" ) ) ? QString() : QStringLiteral( " +type=crs" ) );
+  QgsProjUtils::proj_pj_unique_ptr res( proj_create_crs_to_crs( pContext, "EPSG:4326", projDef.toUtf8(), nullptr ) );
```

Comment:
This won't go very well for non-Earth CRS.
An option would be to instanciate projDef as a PJ* object, and get proj_crs_get_geodetic_crs() from it. You should check that the return of that function is not NULL, and that it is a geographic CRS with proj_get_type().  And finally use proj_normalize_for_visualization() on it, to get a longitude, latitude ordered CRS (you'll need to change the order in proj_trans_generic())
Then you can use proj_create_crs_to_crs_from_pj()

Reply:
@rouault 

Coming back to this one after an (Extended !! :smile:  ) break

How does https://github.com/qgis/QGIS/pull/46443/commits/64ca557415de6a13f6b4a072ab5d964437cbc520 look to you? I've tried to generalize this function in case it's useful elsewhere...

Reply:
comment on naming left in https://github.com/qgis/QGIS/commit/64ca557415de6a13f6b4a072ab5d964437cbc520#r63265883